helix-core/src/main/java/org/apache/helix/controller/rebalancer/CustomRebalancer.java

```diff
@@ -130,14 +130,17 @@ public ResourceAssignment computeBestPossiblePartitionState(ResourceControllerDa
 
     Map<String, LiveInstance> liveInstancesMap = cache.getLiveInstances();
     for (String instance : idealStateMap.keySet()) {
-      boolean notInErrorState =
-          currentStateMap == null || currentStateMap.get(instance) == null
-              || !currentStateMap.get(instance).equals(HelixDefinedState.ERROR.toString());
-
+      boolean notInErrorState = currentStateMap != null
+          && !HelixDefinedState.ERROR.toString().equals(currentStateMap.get(instance));
       boolean enabled = !disabledInstancesForPartition.contains(instance) && isResourceEnabled;
 
-      if (liveInstancesMap.containsKey(instance) && notInErrorState && enabled) {
-        instanceStateMap.put(instance, idealStateMap.get(instance));
+      if (liveInstancesMap.containsKey(instance) && notInErrorState) {
```

Comment:
One more things, even an instance is not live, we should put the bestpossible as OFFLINE too, right?  The current code will remove it from BestPossible.

Reply:
@lei-xia I am not sure. If an instance is not live, we won't be sending it a message anyways. So there's no point in including it in the best possible mapping.

Suppose that instance comes back online later, then that would be a liveInstance change, which would require the cache to be invalidated, forcing the rebalancer to recompute the assignment. 

So it's okay to not include it in the best possible mapping here.